### ISSUE TITLE
startup crash: 'Reactor' object has no attribute 'answer'

### ISSUE THREAD

**Issue by adrian62** [Exa-Networks/exabgp#679]
- Bug Report

##### OS
Ubuntu 16.04LTS

##### VERSION

```
ExaBGP version : 4.0.1-4501ddf3
Python version : 2.7.12 (default, Nov 19 2016, 06:48:10)  [GCC 5.4.0 20160609]
System Uname   : #93-Ubuntu SMP Fri Mar 31 14:07:41 UTC 2017
System MaxInt  : 9223372036854775807
```

##### ENVIRONMENT 

```
[empty]
```

##### CONFIGURATION
<!--- Paste verbatim your configuration file
```
template {
    neighbor GROUP_EXA_AFX_2_MISRR {
        family {
          ipv4 unicast;
          ipv6 nlri-mpls;
        }
        api {
              processes [ receive-routes ];
              receive {
                        parsed;
                        update;
              }
#              neighbor-changes;
        }
    }
    neighbor GROUP_EXA_AFX_2_IAVRR {
        family {
          ipv4 mpls-vpn;
          ipv6 mpls-vpn;
        }
        api {
              processes [ inject-routes ];
              send {
                        update;
              }
              neighbor-changes;
        }
    }
}

process inject-routes {
       # get updates via the other instance's RECV pipe
         run /opt/AFX1.5/uafx.py;
         encoder json;
}
process receive-routes {
       # get updates from the MIS-RR
         run /home/AFX/recRoute.sh /home/AFX/RECV;
         encoder json;
}

neighbor 198.18.0.2{
        description "peer IOX2 vRR";
        inherit GROUP_EXA_AFX_2_IAVRR;
        router-id 10.0.99.11;
        local-address 10.0.99.11;
        local-as 100;
        peer-as 100;
        adj-rib-in false;
        hold-time 180;
        }
neighbor 198.18.0.5{
        description "peer Junos1 MISRR";
        inherit GROUP_EXA_AFX_2_MISRR;
        router-id 10.0.99.11;
        local-address 10.0.99.11;
        local-as 100;
        peer-as 100;
        adj-rib-in false;
        hold-time 180;
        }
```

##### SUMMARY
This works for me on a commit from over a week ago when we teleconferenced:
ExaBGP : (HEAD-388a2c6cbf51de1fbe02ba2202b13ed5762b6f38
But the API ack and the cli don't work in that commit for me, so I pulled this latest one at the time of my writing:
ExaBGP : master-10a80718dbbe3c60a8b8cba675b53160a93c91b5
And this one crashes for me on startup.

##### IMPORTANCE
Can't use API ack, and without it I can't effectively throttle the rate of commands sent on the API.

********************************************************************************
-- Please provide ALL the information below on :
-- https://github.com/Exa-Networks/exabgp/issues
********************************************************************************

ExaBGP version : 4.0.1-4501ddf3
Python version : 2.7.12 (default, Nov 19 2016, 06:48:10)  [GCC 5.4.0 20160609]
System Uname   : #93-Ubuntu SMP Fri Mar 31 14:07:41 UTC 2017
System MaxInt  : 9223372036854775807

-- Traceback

Traceback (most recent call last):
  File "/home/exabgp/lib/exabgp/application/bgp.py", line 418, in <module>
    main()
  File "/home/exabgp/lib/exabgp/application/bgp.py", line 296, in main
    run(env,comment,configurations,root,options["--validate"])
  File "/home/exabgp/lib/exabgp/application/bgp.py", line 362, in run
    was_ok = Reactor(configurations).run(validate,root)
  File "/home/exabgp/lib/exabgp/reactor/loop.py", line 248, in run
    self.api.text(self,service,command)
  File "/home/exabgp/lib/exabgp/reactor/api/__init__.py", line 47, in text
    reactor.answer(service,'error')
AttributeError: 'Reactor' object has no attribute 'answer'

-- Configuration

-- Logging History

********************************************************************************
-- Please provide _ALL_ the information above on :
-- https://github.com/Exa-Networks/exabgp/issues
********************************************************************************

**Comment by thomas-mangin** [comment]
https://github.com/thomas-mangin/exabgp/commit/b5ebb5e2e5965d19008ae195dbed542e743eec5e

**Comment by thomas-mangin** [comment]
oops .. 45344e96

**Comment by adrian62** [comment]
Still having problems with the ack. I added code to read stdin after each command sent on the API. If I turn on ack in the environment, exabgp hangs after ~13000 commands. If I turn off the ack no problems but then I have seen issues with the API being overrun and updates are lost, so I presently have a sleep in the loop to throttle down the rate of commands.
I read the api in the following function. I checked the return to see if there is even an unprintable character, like a linefeed, but it's totally quiet. However the function doesn't block in the read function.

def readFromAPI(**kwargs):
    """read from stdin where the exabgp API responds to our commands; place response in keyword apiAnswer"""
    from sys import stdin
    kwargs['apiAnswer'] = stdin.readline()
    return kwargs

On Thursday, July 13, 2017, 10:28:19 AM PDT, Thomas Mangin <notifications@github.com> wrote:


oops .. 45344e9

**Comment by thomas-mangin** [comment]
Doing correct async IO is hard - I am no exception: I think exabgpcli has something causing a delay I need to investigate.

But the only thing which can stop ExaBGP is a full pipe with the API program. If disabling the messages still fixes the issue then it is indeed the problem.
https://github.com/Exa-Networks/exabgp/blob/master/etc/exabgp/example-api-program.run

I am not sure how the command could be lost. They are queued on receipt and then processed.

**Comment by adrian62** [comment]
Understood, in my early career I worked on hardware... async communication.
Would you consider introducing the ability to interface processes with exa using fifos  (like you did with the cli :) in lieu of i/o redirection of the standard streams? In fact, if I understood correctly, I could use the cli to interface to my process to issue the API commands. but I cannot get it to work, even with the more recent commits:
ubuntu@afx1:/opt/AFX1.5$ sudo -u nobody /home/exabgp/bin/exabgpcli help
could not read answer to ExaBGPubuntu@afx1:/opt/AFX1.5$


On Thursday, July 13, 2017, 12:37:31 PM PDT, Thomas Mangin <notifications@github.com> wrote:


Doing correct async IO is hard - I am no exception: I think exabgpcli has something causing a delay I need to investigate.

But the only thing which can stop ExaBGP is a full pipe with the API program. If disabling the messages still fixes the issue then it is indeed the problem.
https://github.com/Exa-Networks/exabgp/blob/master/etc/exabgp/example-api-program.run

I am not sure how the command could be lost. They are queued on receipt and then processed.

**Comment by thomas-mangin** [comment]
For the command being lost: I know what happens. Incredible what a good night sleep can do to your brain :-)

If you do a print (and not a file write). you are not told how many bytes were written. If you go too fast, you will fill the pipe with data, and following print are "lost".
using sys.stdout.write would tell you that 0 (or more) bytes were written and any value under the string size mean you have to retry with the data not transmitted.

If you look the code I propose as example to connect to ExaBGP, it does it:
https://github.com/Exa-Networks/exabgp/blob/master/etc/exabgp/example-api-program.run#L86

**Comment by thomas-mangin** [comment]
Just because you nearly asked:
320c4d63

So you can use the same code ExaBGP uses for the cli for your own program using pipes.
Just start it in the command line with something like
`env exabgp_api_pipename=control` and create two fifo control.in and control.out in any of the `run` folders scanned ( at least `/var/run` , `'/var/run/exabgp`, `/var/run/<uid>` and the same without `/var` and all this in the local exabgp folder)

**Comment by adrian62** [comment]
I relate to that, I get answers overnite ;)
I have implemented the write portion only (I still have ack=false, therefore no reading) using your async I/O code. The situation didn't improve. I added logging to the code to report any occurrences when the buffer is only partially written on the stream. No hits so I'm still investigating.  
BTW, in your write method, if the os.write call only wrote part of the buffer but not all, local variable 'left' will contain the remainder that wasn't written. However upon return we get 'not not left' which is True if there was a remainder and False if the entire buffer is written. But the information about how much of the data was written, and the data itself is lost. I would think that we want to retry writing just the remainder, and not the entire buffer which would cause the portion that was initially written by os.write to be repeated.
Haven't tried the api over pipes yet, but thanks for being so forthcoming and offering answers ahead of my questions!
